From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: IBM Codedrop 2006-01-09 - linux-2.6.15-s390-base-13-october2005.diff

[patch] multicast notifier.

s390 network driver change:
 - Add option NETIF_F_MC_ALL to get notified for all multicast
   address additions/removals.

Acked-by: Hannes Reinecke <hare@suse.de>

---

 drivers/s390/net/lcs.c       |   16 ++++++++--------
 drivers/s390/net/qeth_main.c |   10 ++++++----
 include/linux/netdevice.h    |    3 +++
 net/core/dev_mcast.c         |    5 +++++
 4 files changed, 22 insertions(+), 12 deletions(-)

diff -urpN linux-2.6/drivers/s390/net/lcs.c linux-2.6-patched/drivers/s390/net/lcs.c
--- linux-2.6/drivers/s390/net/lcs.c	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-patched/drivers/s390/net/lcs.c	2006-01-09 15:25:40.000000000 +0100
@@ -11,7 +11,7 @@
  *			  Frank Pavlic (fpavlic@de.ibm.com) and
  *		 	  Martin Schwidefsky <schwidefsky@de.ibm.com>
  *
- *    $Revision: 1.99 $	 $Date: 2005/05/11 08:10:17 $
+ *    $Revision: 1.102 $	 $Date: 2006/01/04 12:21:29 $
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -59,7 +59,7 @@
 /**
  * initialization string for output
  */
-#define VERSION_LCS_C  "$Revision: 1.99 $"
+#define VERSION_LCS_C  "$Revision: 1.102 $"
 
 static char version[] __initdata = "LCS driver ("VERSION_LCS_C "/" VERSION_LCS_H ")";
 static char debug_buffer[255];
@@ -101,9 +101,9 @@ lcs_register_debug_facility(void)
 		return -ENOMEM;
 	}
 	debug_register_view(lcs_dbf_setup, &debug_hex_ascii_view);
-	debug_set_level(lcs_dbf_setup, 4);
+	debug_set_level(lcs_dbf_setup, 2);
 	debug_register_view(lcs_dbf_trace, &debug_hex_ascii_view);
-	debug_set_level(lcs_dbf_trace, 4);
+	debug_set_level(lcs_dbf_trace, 2);
 	return 0;
 }
 
@@ -1295,9 +1295,8 @@ lcs_set_multicast_list(struct net_device
         LCS_DBF_TEXT(4, trace, "setmulti");
         card = (struct lcs_card *) dev->priv;
 
-        if (!lcs_set_thread_start_bit(card, LCS_SET_MC_THREAD)) {
+        if (!lcs_set_thread_start_bit(card, LCS_SET_MC_THREAD)) 
 		schedule_work(&card->kernel_thread_starter);
-	}
 }
 
 #endif /* CONFIG_IP_MULTICAST */
@@ -2197,8 +2196,10 @@ lcs_new_device(struct ccwgroup_device *c
 	SET_MODULE_OWNER(dev);
 	memcpy(card->dev->dev_addr, card->mac, LCS_MAC_LENGTH);
 #ifdef CONFIG_IP_MULTICAST
-	if (!lcs_check_multicast_support(card))
+	if (!lcs_check_multicast_support(card)) {
 		card->dev->set_multicast_list = lcs_set_multicast_list;
+		card->dev->features |= NETIF_F_MC_ALL; 
+	}
 #endif
 netdev_out:
 	lcs_set_allowed_threads(card,0xffffffff);
@@ -2322,7 +2323,6 @@ __init lcs_init_module(void)
 		PRINT_ERR("Initialization failed\n");
 		return rc;
 	}
-
 	return 0;
 }
 
diff -urpN linux-2.6/drivers/s390/net/qeth_main.c linux-2.6-patched/drivers/s390/net/qeth_main.c
--- linux-2.6/drivers/s390/net/qeth_main.c	2006-01-09 15:25:37.000000000 +0100
+++ linux-2.6-patched/drivers/s390/net/qeth_main.c	2006-01-09 15:25:40.000000000 +0100
@@ -1,6 +1,6 @@
 /*
  *
- * linux/drivers/s390/net/qeth_main.c ($Revision: 1.251 $)
+ * linux/drivers/s390/net/qeth_main.c ($Revision: 1.252.2.1 $)
  *
  * Linux on zSeries OSA Express and HiperSockets support
  *
@@ -12,7 +12,7 @@
  *			  Frank Pavlic (fpavlic@de.ibm.com) and
  *		 	  Thomas Spatzier <tspat@de.ibm.com>
  *
- *    $Revision: 1.251 $	 $Date: 2005/05/04 20:19:18 $
+ *    $Revision: 1.252.2.1 $	 $Date: 2006/01/04 17:59:20 $
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -73,7 +73,7 @@
 #include "qeth_eddp.h"
 #include "qeth_tso.h"
 
-#define VERSION_QETH_C "$Revision: 1.251 $"
+#define VERSION_QETH_C "$Revision: 1.252.2.1 $"
 static const char *version = "qeth S/390 OSA-Express driver";
 
 /**
@@ -7069,6 +7069,7 @@ qeth_start_ipa_multicast(struct qeth_car
 	} else {
 		PRINT_INFO("Multicast enabled\n");
 		card->dev->flags |= IFF_MULTICAST;
+		card->dev->features |= NETIF_F_MC_ALL;
 	}
 	return rc;
 }
@@ -7460,7 +7461,8 @@ qeth_softsetup_card(struct qeth_card *ca
 		card->dev->features |=
 			NETIF_F_HW_VLAN_FILTER |
 			NETIF_F_HW_VLAN_TX |
-			NETIF_F_HW_VLAN_RX;
+			NETIF_F_HW_VLAN_RX |
+			NETIF_F_MC_ALL;
 		card->dev->flags|=IFF_MULTICAST|IFF_BROADCAST;
 		card->info.broadcast_capable=1;
 		if ((rc = qeth_layer2_initialize(card))) {
diff -urpN linux-2.6/include/linux/netdevice.h linux-2.6-patched/include/linux/netdevice.h
--- linux-2.6/include/linux/netdevice.h	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-patched/include/linux/netdevice.h	2006-01-09 15:25:40.000000000 +0100
@@ -309,6 +309,9 @@ struct net_device
 #define NETIF_F_TSO		2048	/* Can offload TCP/IP segmentation */
 #define NETIF_F_LLTX		4096	/* LockLess TX */
 #define NETIF_F_UFO             8192    /* Can offload UDP Large Send*/
+#define NETIF_F_MC_ALL		16384   /* trigger driver on every multicast
+	                                 * address been added/deleted
+				         */
 
 	struct net_device	*next_sched;
 
diff -urpN linux-2.6/net/core/dev_mcast.c linux-2.6-patched/net/core/dev_mcast.c
--- linux-2.6/net/core/dev_mcast.c	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6-patched/net/core/dev_mcast.c	2006-01-09 15:25:40.000000000 +0100
@@ -145,6 +145,8 @@ int dev_mc_delete(struct net_device *dev
 	}
 	err = -ENOENT;
 done:
+	if (dev->features & NETIF_F_MC_ALL)
+		__dev_mc_upload(dev);
 	spin_unlock_bh(&dev->xmit_lock);
 	return err;
 }
@@ -193,6 +195,9 @@ int dev_mc_add(struct net_device *dev, v
 	return 0;
 
 done:
+	if (dev->features & NETIF_F_MC_ALL)
+		__dev_mc_upload(dev);
+
 	spin_unlock_bh(&dev->xmit_lock);
 	kfree(dmi1);
 	return err;
